### DISCUSSION TITLE
[Text] table

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-better-notes#221] [Note Templates]
Type: Text

Description: Render a table in your note according to the input size.

See https://github.com/windingwind/zotero-better-notes/blob/master/TemplateDoc.md#table

![image](https://user-images.githubusercontent.com/33902321/184603621-dcb74d0c-8c73-4026-99e5-b84b2af27806.png)

```js
${(()=>{
const size = prompt("Table Size(row*column):", "4*3");
if(!size){
return "";
}
const row = Number(size.split("*")[0]);
const col = Number(size.split("*")[1]);
if(!row || !col){
return "";
}
const makeCell = ()=>"<td>\n</td>";
const makeRow = ()=>`<tr>${[...Array(col).keys()].map(makeCell).join("\n")}</tr>`;
return `<table>
${[...Array(row).keys()].map(makeRow).join("\n")}
</table>`;
})()}
```

**Comment by Cucu-zhao** [comment]
Excuse me could tell me why I can't type Chinese text in the cell？

**Comment by Philip-Cao-9527** [comment]
I'm wondering why I  cannot type any words in the table generated by the code whether in Chinese or English

**Reply by kuyree** [reply]
主笔记插入表格，可以编辑打字；item笔记里插入表格，不能打字，只能复制黏贴

**Reply by Philip-Cao-9527** [reply]
> 主笔记插入表格，可以编辑打字；item笔记里插入表格，不能打字，只能复制黏贴
是的，我遇到了同样的问题，item笔记里面插入偶尔可以打字，但是容易卡顿，必须要刷新之后才能输入，这样子太麻烦了

**Reply by baishn1289** [reply]
表格模板我遇到了相同的问题，没法输入任何信息
兄弟你解决了吗

**Reply by LeonaGegeCui** [reply]
Ctrl+左键单击 应该可以激活单元格输入

**Comment by censalata** [comment]
This feature is not working for me. It returns a Javascript Application error: Template [Text] table Error: SyntaxError: "" string literal contains an unescaped line break

**Reply by lllaterOn** [reply]
我有同样的问题

**Reply by CrushDemo01** [reply]
Me too.

**Reply by CrushDemo01** [reply]
I have rewritten some of the content, which is useful to me.
```
${(()=>{

const size = prompt("Table Size(rowcolumn):", "43");

if (!size) {
return "";
}

const row = Number(size.split("")[0]);
const col = Number(size.split("")[1]);

const generateTable = (row, col) => {
const makeHeadCell = () => "";
const makeHead = () => <tr>${[...Array(col).keys()].map(makeHeadCell).join("")}</tr>;
const makeCell = () => "";
const makeRow = () => <tr>${[...Array(col).keys()].map(makeCell).join("")}</tr>;
return <table> <thead>${makeHead()}</thead> ${"<tbody>" + [...Array(row - 1).keys()].map(makeRow).join("") + "</tbody>"} </table>;
};

const tableHtml = generateTable(row, col);
return tableHtml;
})()}
```
Sorry, some of the codes above are missing. Try replying code below.

**Reply by baishn1289** [reply]
It returns a JavaScript Application: Template [Text] table Error: SyntaxError: expected expression, got '<'

**Reply by CrushDemo01** [reply]
Sorry, some codes are missing. Try this.
```
${(()=>{
    const size = prompt("Table Size(row*column):", "4*3");

    if (!size) {
      return "";
    }
    
    const row = Number(size.split("*")[0]);
    const col = Number(size.split("*")[1]);
    
    const generateTable = (row, col) => {
      const makeHeadCell = () => "<th></th>";
      const makeHead = () => `<tr>${[...Array(col).keys()].map(makeHeadCell).join("")}</tr>`;
      const makeCell = () => "<td></td>";
      const makeRow = () => `<tr>${[...Array(col).keys()].map(makeCell).join("")}</tr>`;
      return `<table>
                <thead>${makeHead()}</thead>
                ${"<tbody>" + [...Array(row - 1).keys()].map(makeRow).join("") + "</tbody>"}
              </table>`;
    };
    
    const tableHtml = generateTable(row, col);
    return tableHtml;
})()}
```

**Reply by baishn1289** [reply]
I found that it did generate a table, but I can't type any words either in English or Germany.
Is there still something wrong?

**Reply by CrushDemo01** [reply]
yes. 跟这个情况一样[[kuyree](https://github.com/windingwind/zotero-better-notes/discussions/221#discussioncomment-4836712)
只有在主笔记下才能填充内容，其余情况只能复制进去内容。
我也不清楚为什么这样

**Reply by windingwind** [reply]
Please comment here: https://forums.zotero.org/discussion/104653/bug-cannot-move-cursor-into-table-in-note-editor#latest

to let the Zotero team notice this bug.

**Comment by windingwind** [comment]
Thanks you all for your efforts to making it work, however: 

1. The coming new version 1.0.0+ supports markdown template, thus most old templates using HTML should be upgraded to using markdown. HTML templates can still work, but they are more difficult to be maintained.
2. Tables in the note editor sometimes cannot display cursor correctly. This might be a bug with Zotero. You could use right-click to move cursor into the table.

**Reply by CrushDemo01** [reply]
Thank you for your contribution. Looking forward to better support for markdown!

**Comment by chen1573** [comment]
我保存这个模板时提示
`Template [Text] table Error: ReferenceError: prompt is not defined`
该怎么解决呢

**Reply by tina140119** [reply]
> 我保存这个模板时提示 `Template [Text] table Error: ReferenceError: prompt is not defined` 该怎么解决呢

我也遇到了这个问题。so，你可以自己写一个代码来解决这个问题。除此之外，就是各种无法在自己建立的table里打字= =、

**Reply by drggboy** [reply]
将代码中 prompt 改为 Zotero.getMainWindow().prompt 后成功显示：
~~~
${(()=>{
    const size = Zotero.getMainWindow().prompt("Table Size(row*column):", "4*3");

    if (!size) {
      return "";
    }
    
    const row = Number(size.split("*")[0]);
    const col = Number(size.split("*")[1]);
    
    const generateTable = (row, col) => {
      const makeHeadCell = () => "<th></th>";
      const makeHead = () => `<tr>${[...Array(col).keys()].map(makeHeadCell).join("")}</tr>`;
      const makeCell = () => "<td></td>";
      const makeRow = () => `<tr>${[...Array(col).keys()].map(makeCell).join("")}</tr>`;
      return `<table>
                <thead>${makeHead()}</thead>
                ${"<tbody>" + [...Array(row - 1).keys()].map(makeRow).join("") + "</tbody>"}
              </table>`;
    };
    
    const tableHtml = generateTable(row, col);
    return tableHtml;
})()}
~~~
参考：[[Bug] 大部分模板无法使用，如prompt not defined, 以及 topItem not defined等问题 · Issue #476 · windingwind/zotero-better-notes](https://github.com/windingwind/zotero-better-notes/issues/476)

**Comment by Lance-Tin** [comment]
创建好之后咋删除不了了

**Reply by iqeix** [reply]
你可以选中表格，然后右键->删除表格

**Comment by keyanxiaobai-xiaopei** [comment]
您好！请问为什么我在创建模板时，预览框显示的是“ReferenceError: prompt is not defined ”，谢谢

**Reply by windingwind** [reply]
见https://github.com/windingwind/zotero-better-notes/discussions/221#discussioncomment-6147185
